feat(nx-plugin): add NX Plugin with OpenAPI client generator and executor

seriouslag opened the pull request:
**Update June 2026**
I am keeping this up to date at [seriouslag/openapi-ts-nx-plugin](https://github.com/seriouslag/openapi-ts-nx-plugin) and plan to do so for the foreseeable future.
Any feature requests or notes post there. 




----------

This NX plugin will generate an NX package from an OpenAPI spec file and run the openapi-ts on the spec.

This is useful for NX projects.
Also very useful to let AI agents know how to use the plugin to automate setting up client code for an NX project.




TODO:
- [ ] Add support for NX Inferred Tasks
- [x] Add NX generator to generate library code and generate client code based off an openapi spec file
- [x] Add NX executor to check the provided spec file for changes and update client code
- [x] Accept openapi spec input, dereference, and bundle the spec file using functions from openapi-ts
- [x] Setup tests using vitest
- [x] Add an executor to update the API spec file and regenerate the client code
- [x] Get build working (facing issues with integrating into monorepo)
- [x] Make plugins configurable and add the ability to add additional openapi-ts plugins
- [x] Support additional client types
- [X] Configure the project linking for the executor if the API spec is in another project
- [X] Auto-symlink the packages if in workspace mode, using the native package manager
- [X] Support NX 21 continuous pipelines
- [ ] List the NX plugin in NX plugin repository

Stretch in this PR:
- [ ] Find or build an OpenAPI diff package that supports OpenAPI 2 through 3.1. We are using [api-smart-diff](https://www.npmjs.com/package/api-smart-diff). There haven't been many updates, and there's no Swagger 2 support. We are getting around this by upgrading the Swagger files to OpenAPI v3 files before diffing.
- [ ] Support different test frameworks (missing jest)
- [ ] Generate linting support (eslint, etc)

Questions:
Q: How to handle updates to the OpenAPI spec file?
A: We have added an executor and a call to each generated project. Each project can run the executor to update the spec file and regenerate the client code.

Q: How do we list the plugin in the NX plugin repository
A: [https://nx.dev/extending-nx/recipes/publish-plugin#list-your-nx-plugin](https://nx.dev/extending-nx/recipes/publish-plugin#list-your-nx-plugin)

Fixes: #1910 

Related PRs:
- [Update Vitest](https://github.com/hey-api/openapi-ts/pull/1934)
- [Export common functions to dogfood capabilities](https://github.com/hey-api/openapi-ts/pull/1934)

seriouslag committed:
feat(nx-plugin): add nx plugin to generate and update client code

seriouslag committed:
feat(config): add exports field and support for CommonJS format in Vite base config

seriouslag committed:
fix(turbo.json): enable caching for build task

seriouslag committed:
fix(nx-plugin): update build output

seriouslag committed:
refactor(nx-plugin): update schema prompts and descriptions for clarity in update-api and openapi-client

seriouslag committed:
chore(nx-plugin): remove ts-node dependency and update JSON file copying script to js from ts

seriouslag committed:
feat(vitest): add setup files configuration to Vitest base config

seriouslag committed:
feat(turbo.json): add inputs configuration for improved build process

seriouslag committed:
feat(client): update example to use actual code from the spec to avoid type error

seriouslag committed:
feat(config): update setup file reference in Vitest base config to use module path

seriouslag committed:
feat(turbo.json): add build task configuration for Nuxt project

seriouslag commented on examples/openapi-ts-nuxt/turbo.json:
```diff
@@ -0,0 +1,11 @@
+{
```
Added this file to cache builds

seriouslag commented on examples/openapi-ts-sample/src/App.tsx:
```diff
@@ -9,14 +9,14 @@ import {
 function App() {
   const onClick = async () => {
-    postFoo({
+    addPet({
```
Changed this file to use actual generated code, I was getting type errors how it was

mrlubos replied:
Sample is meant to be a playground for local development and testing. I thought I ignored it in ESLint?

seriouslag replied:
Maybe was an issue with my IDE, why not use actual code instead of adding broken imports and ignoring it with comments?

mrlubos replied:
I don't remember, I was probably testing something and didn't clean it up. It's similar to the sample snapshot script except it contains a whole app with UI. The idea is to be able to test drive the output since snapshots are never actually executed. Obviously should be way more polished!

seriouslag commented on examples/openapi-ts-sample/src/client/client.gen.ts:
```diff
@@ -1,7 +1,32 @@
 // This file is auto-generated by @hey-api/openapi-ts
```
Ran the code gen

seriouslag commented on packages/config-vite-base/package.json:
```diff
@@ -10,6 +10,19 @@
   "scripts": {
     "build": "tsup"
   },
+  "exports": {
```
Was facing issues with our new configs, adding the exports and adding cjs fixed the issues,

seriouslag commented on packages/config-vite-base/package.json:
```diff
@@ -10,6 +10,19 @@
+        "default": "./dist/index.cjs"
+      }
+    },
+    "./setup": "./src/setup.ts"
```
Exporting the setup file as a direct source file for other vitests to use as a setup file.

seriouslag commented on packages/config-vite-base/src/setup.ts:
```diff
@@ -0,0 +1,5 @@
+// FAIL LOUDLY on unhandled promise rejections / errors
```
Adding this file because vitest fails if an error is uncaught, but no trace is given to the logs. This change adds the trace. This will help future debugging.

seriouslag commented on packages/config-vite-base/src/vitest.base.config.ts:
```diff
@@ -25,6 +25,7 @@ export function createVitestConfig(
         },
       },
       root,
+      setupFiles: ['@config/vite-base/setup'],
```
Adding our new vitest setup file with the new export path

seriouslag commented on turbo.json:
```diff
@@ -2,7 +2,7 @@
   "tasks": {
     "build": {
-      "cache": false,
+      "cache": true,
```
Enabled caching for quicker local dev

seriouslag committed:
fix(nx-plugin): fix plugin build output and enhance caching

- fix typo in vitest.config output
- change vitest.config to mts for better type support
- list vite and vitest as a dep if enabled
- change vitest to single fork runner
- change build to only output cjs to support nx plugins
- add caching and nx linking if spec file is in another project

seriouslag committed:
refactor(nx-plugin): enhance plugin handling and directory management

- Replace direct mkdir calls with makeDir utility for consistent directory creation.
- Update plugin handling in client generation to support additional plugins.
- Modify tests and schemas to reflect changes in plugin structure and options.
- Improve error handling and logging in updateApi executor.
- Add formatting functionality for generated files and update README for clarity.

seriouslag committed:
feat(nx-plugin): enhance OpenAPI spec comparison functionality

- Replace the existing openapi-diff dependency with api-smart-diff for improved comparison accuracy.
- Add comprehensive test cases to validate spec comparison logic across various scenarios.

seriouslag committed:
chore(dependencies): downgrade find-my-way to version 9.2.0 and add overrides in package.json files to support node 18

seriouslag committed:
feat(nx-plugin): add base tsconfig options for improved configuration management

- Introduce `baseTsConfigName` and `baseTsConfigPath` options in the OpenAPI client generator and update executor schemas.
- Enhance README documentation to reflect new options and their usage.
- Implement `getBaseTsConfigPath` utility to resolve tsconfig paths based on provided options.
- Update tests to cover new tsconfig path handling logic.

seriouslag committed:
feat(nx-plugin): enhance OpenAPI and Swagger spec comparison utilities

- Introduce functions to standardize OpenAPI and Swagger specs, including version detection and conversion.
- Implement `removeExamples` function to clean up specs before comparison, ensuring accurate diff results.
- Add comprehensive tests for new functionality, covering various scenarios for OpenAPI 3.1 and Swagger 2.0 specs.
- Update existing comparison logic to utilize standardized specs for improved accuracy.

Le0Developer committed:
draft typeid support

seriouslag committed:
feat(nx-plugin): update to support NX 21 continuous pipelines

seriouslag committed:
major: add node 24 in CI, remove node 18 as a supported target

seriouslag committed:
refactor(nx-plugin): improve file formatting utilities

Commit:
fix(zod): add metadata option

mrlubos committed:
Merge pull request #2163 from hey-api/fix/zod-metadata

fix(zod): add metadata option

Commit:
fix(client): do not serialize path param name in url

mrlubos committed:
Merge pull request #2165 from hey-api/fix/client-path-param-date

fix(client): do not serialize path param name in url

Commit:
fix(zod): support tuple types

mrlubos committed:
Merge pull request #2166 from hey-api/fix/zod-tuple

fix(zod): support tuple types

Commit:
fix(tanstack-query): add name builder options for all generated artifacts

mrlubos committed:
Merge pull request #2167 from hey-api/fix/tanstack-query-name-builder

fix(tanstack-query): add name builder options for all generated artifacts

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2164 from hey-api/changeset-release/main

Version Packages

Commit:
fix: export default pagination keywords

mrlubos committed:
Merge pull request #2170 from hey-api/fix/export-default-pagination-keywords

fix: export default pagination keywords

Commit:
feat: remove support for client packages

Commit:
fix: respect logs setting if initialization fails

Commit:
feat: bundle fetch client

Commit:
feat: bundle axios client

Commit:
feat: bundle next client

Commit:
docs: update client docs

Commit:
feat: bundle nuxt client

Commit:
chore: update custom local client

Commit:
chore: update custom third-party client

Commit:
test: update snapshots

Commit:
chore: update Angular example

Commit:
docs: add changesets

Commit:
chore: fix tests

mrlubos committed:
Merge pull request #2172 from hey-api/feat/client-bundle

feat: remove support for client packages

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2171 from hey-api/changeset-release/main

Version Packages

Commit:
docs: add changesets

mrlubos committed:
Merge pull request #2175 from hey-api/feat/nuxt-module

docs: add changesets

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2176 from hey-api/changeset-release/main

Version Packages

Commit:
docs: add code blocks

mrlubos committed:
Merge pull request #2177 from hey-api/docs/readme-code-block

docs: add code blocks

mrlubos committed:
refactor: namespace plugin configs

mrlubos committed:
Merge pull request #2180 from hey-api/refactor/plugin-config

refactor: namespace plugin configs

mrlubos committed:
docs: add client section on custom implementation

mrlubos committed:
Merge pull request #2181 from hey-api/docs/client-custom-implementation

docs: add client section on custom implementation

mrlubos committed:
refactor: clean up plugin types a bit

mrlubos committed:
Merge pull request #2182 from hey-api/refactor/plugin-types

mrlubos committed:
refactor: clean up more plugin types

Daschi1 committed:
fix(valibot): use v.isoTimestamp instead of v.isoDateTime for OpenAPI's date-time

mrlubos committed:
Merge pull request #2185 from hey-api/refactor/plugin-api

refactor: clean up more plugin types

Daschi2 committed:
test(valibot): update test snapshot files

Daschi1 committed:
Merge branch 'main' into patch-1

mrlubos committed:
Create odd-yaks-love.md

mrlubos committed:
Merge pull request #2192 from Daschi1/patch-1

mrlubos committed:
chore: add resolved config types for Zod

mrlubos committed:
feat: add customization options for Zod plugin

mrlubos committed:
feat: add parameters to Zod request schemas

mrlubos committed:
docs: update Zod documentation

mrlubos committed:
fix(parser): add format to deduplcation strategy

mrlubos committed:
Merge branch 'main' into fix/plugin-transform

mrlubos committed:
Merge pull request #2201 from hey-api/fix/plugin-transform

Zod: add `name` builders and `case` options

mrlubos committed:
docs: bump stats

mrlubos committed:
Merge pull request #2208 from hey-api/docs/stats-bump

docs: bump stats

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2193 from hey-api/changeset-release/main

Version Packages

mrlubos committed:
fix(cli): correctly handle watch mode

mrlubos committed:
Merge pull request #2210 from hey-api/fix/bin-watch

fix(cli): correctly handle watch mode

mrlubos committed:
fix(validators): handle additional properties object when no other properties are defined

mrlubos committed:
Merge pull request #2213 from hey-api/fix/validators-additional-properties-object

fix(validators): handle additional properties object when no other properties are defined

mrlubos committed:
refactor(parser): replace plugin.subscribe with plugin.walk

mrlubos committed:
docs: add migration for plugin.forEach

mrlubos committed:
Merge pull request #2215 from hey-api/refactor/plugin-walk

refactor(parser): replace plugin.subscribe with plugin.walk

mrlubos committed:
fix(parser): add meta and version options to input.patch

mrlubos committed:
Merge pull request #2216 from hey-api/fix/parser-patch-meta

fix(parser): add meta and version options to input.patch

mrlubos committed:
feat(tanstack-query): add name and case options

mrlubos committed:
docs: update tanstack query documentation

mrlubos committed:
Merge pull request #2218 from hey-api/feat/tanstack-query-api

feat(tanstack-query): add name and case options

mrlubos committed:
fix: make output pass stricter tsconfig configurations

mrlubos committed:
Merge pull request #2219 from hey-api/fix/output-types

fix: make output pass stricter tsconfig configurations

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2211 from hey-api/changeset-release/main

mrlubos committed:
fix(parser): prefer JSON media type

mrlubos committed:
Merge pull request #2221 from hey-api/fix/axios-client-prefer-json

fix(parser): prefer JSON media type

mrlubos committed:
feat(valibot): single schema per request

mrlubos committed:
Merge pull request #2226 from hey-api/feat/valibot-config

feat(valibot): single schema per request

github-actions[bot] committed:
Version Packages

mrlubos committed:
feat(sdk): update validator config types

mrlubos committed:
Merge pull request #2222 from hey-api/changeset-release/main

Version Packages

mrlubos committed:
feat(plugin): add api field

mrlubos committed:
feat(client): add requestValidator option

mrlubos committed:
refactor(plugin): return plugin interface from defineplugin

mrlubos committed:
test: update snapshots

mrlubos committed:
docs: update validators pages

mrlubos committed:
Merge branch 'main' into feat/sdk-validator-request

mrlubos committed:
docs: fix dead link error

mrlubos committed:
Merge pull request #2227 from hey-api/feat/sdk-validator-request

feat(sdk): add request validators

mrlubos committed:
docs: add Cella sponsorship

mrlubos committed:
Merge pull request #2241 from hey-api/docs/sponsor-cella

docs: add Cella sponsorship

mrlubos committed:
docs: correctly symlink readme

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2242 from hey-api/docs/readme-symlink

docs: correctly symlink readme

mrlubos committed:
Merge pull request #2240 from hey-api/changeset-release/main

Version Packages

mrlubos committed:
refactor(typescript): move enum options under enums

mrlubos committed:
refactor(typescript): rename readable and writable name builders

mrlubos committed:
refactor(parser): move parser config to a separate field

mrlubos committed:
refactor(typescript): move exportInlineEnums to parser options

mrlubos committed:
feat(parser): add enums transforms

mrlubos committed:
refactor(typescript): move readWrite settings to parser options

mrlubos committed:
feat(parser): move transforms to a separate pass

mrlubos committed:
docs: bump stats

mrlubos committed:
Merge pull request #2256 from hey-api/docs/stats-bump

docs: bump stats

himself65 committed:
fix: add `undefined` type in key map

mrlubos committed:
Merge branch 'main' into feat/typescript-config

mrlubos committed:
test: bump threshold

mrlubos committed:
docs: add parser documentation

mrlubos committed:
Merge pull request #2246 from hey-api/feat/typescript-config

Add `parser` options 👨‍🔧

mrlubos committed:
docs: clean up changesets

mrlubos committed:
Merge pull request #2273 from hey-api/docs/v0-78-0-changesets

docs: clean up changesets

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2272 from hey-api/changeset-release/main

Version Packages

mrlubos committed:
fix(valibot): properly handle array minLength and maxLength

mrlubos committed:
Merge pull request #2275 from hey-api/fix/valibot-array-min-max-length

fix(valibot): properly handle array minLength and maxLength

mrlubos committed:
fix(clients): annotate serializer return types

mrlubos committed:
Merge pull request #2277 from hey-api/fix/client-serializer-return-type

fix(clients): annotate serializer return types

mrlubos committed:
fix(typescript): handle additionalProperties in propertyNames

mrlubos committed:
Merge pull request #2279 from hey-api/fix/typescript-property-names

fix(typescript): handle additionalProperties in propertyNames

mrlubos committed:
fix(zod): add dates.offset option

mrlubos committed:
Merge pull request #2280 from hey-api/fix/zod-dates-offset

fix(zod): add dates.offset option

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2276 from hey-api/changeset-release/main

Version Packages

j-ibarra committed:
custom transformers

mrlubos committed:
fix(typescript): add name option

mrlubos committed:
Create giant-carrots-change.md

mrlubos committed:
Merge pull request #2281 from j-ibarra/main

custom transformers

mrlubos committed:
refactor(transformers): rename types

mrlubos committed:
Merge pull request #2285 from hey-api/refactor/transformers-types

refactor(transformers): rename types

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2283 from hey-api/changeset-release/main

Version Packages

btmnk committed:
refactor(client-fetch): improves client-fetch types to allow usage of ts-reset

btmnk committed:
refactor(client-fetch): updates test snapshots

mrlubos committed:
Create tall-papayas-pay.md

mrlubos committed:
Merge pull request #2290 from btmnk/improves-client-fetch-types

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2292 from hey-api/changeset-release/main

mrlubos committed:
Merge branch 'main' into fix/typescript-name

mrlubos committed:
test: fix tests

mrlubos committed:
Merge pull request #2284 from hey-api/fix/typescript-name

fix(typescript): add name option

mrlubos committed:
docs: fix incorrect references in typescript

mrlubos committed:
Merge pull request #2298 from hey-api/docs/typescript-fix

docs: fix incorrect references in typescript

mrlubos committed:
chore(sdk): initial implementation for custom params

mrlubos committed:
docs: update changesets

mrlubos committed:
Merge pull request #2322 from hey-api/docs/changeset

docs: update changesets

mrlubos committed:
chore: update turbo

mrlubos committed:
Merge branch 'main' into chore/sdk-params-progress

mrlubos committed:
chore: update snapshots

mrlubos committed:
Merge pull request #2310 from hey-api/chore/sdk-params-progress

chore(sdk): initial implementation for custom params

mrlubos committed:
feat(zod): add types option

mrlubos committed:
Merge pull request #2327 from hey-api/feat/zod-types

feat(zod): add types option

mrlubos committed:
docs: add community spotlight

mrlubos committed:
Merge pull request #2329 from hey-api/docs/community

docs: add community spotlight

mrlubos committed:
docs: small fixes to community updates

mrlubos committed:
Merge pull request #2330 from hey-api/docs/community-fix

docs: small fixes to community updates

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2299 from hey-api/changeset-release/main

Version Packages

mrlubos committed:
fix(client): update Axios header types

mrlubos committed:
Merge pull request #2331 from hey-api/fix/axios-client-types

fix(client): update Axios header types

mrlubos committed:
fix(tanstack-query): set query key base url from options if defined

mrlubos committed:
Merge pull request #2263 from himself65/fix

fix: add `undefined` type in key map

mrlubos committed:
Merge pull request #2333 from hey-api/fix/tanstack-query-key-base-url

fix(tanstack-query): set query key base url from options if defined

mrlubos committed:
fix: improve handlebars types for jsr compliance

mrlubos committed:
Merge pull request #2334 from hey-api/fix/handlebars-jsr

fix: improve handlebars types for jsr compliance

mrlubos committed:
chore: check Node version

mrlubos committed:
Merge pull request #2335 from hey-api/chore/check-node-version

chore: check Node version

mrlubos committed:
fix: respect NO_INTERACTIVE and NO_INTERACTION environment variables

mrlubos committed:
Merge pull request #2336 from hey-api/fix/tty

fix: respect NO_INTERACTIVE and NO_INTERACTION environment variables

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2332 from hey-api/changeset-release/main

Version Packages

github-actions[bot] committed:
chore(docs): update contributors list

mrlubos committed:
docs: add Zod 4 page

alexvuka1 committed:
Merge branch 'main' into fix-operation-id-casing

mrlubos committed:
Create witty-dolls-applaud.md

mrlubos committed:
Merge pull request #2041 from alexvuka1/fix-operation-id-casing

fix: operationToId should convert case based on config

mrlubos committed:
Create clean-kiwis-worry.md

mrlubos committed:
Merge pull request #2034 from Le0Developer/feat/typeids

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2342 from hey-api/changeset-release/main

Version Packages

github-actions[bot] committed:
chore(docs): update contributors list

mrlubos committed:
chore: add Zod compatibility version

mrlubos committed:
chore: add Zod module detection

mrlubos committed:
chore: Zod 4 plugin

mrlubos committed:
chore: Zod Mini plugin

mrlubos committed:
test: update snapshots

mrlubos committed:
test: add Zod 3 snapshots

mrlubos committed:
Merge branch 'main' of https://github.com/hey-api/openapi-ts into feat/zod-4

mrlubos committed:
test: clean up Zod 3 tests

mrlubos committed:
test: update snapshots

mrlubos committed:
docs: add supported version labels to plugins

mrlubos committed:
docs: add documentation for Zod 4 and Zod Mini

mrlubos committed:
Merge pull request #2341 from hey-api/feat/zod-4

docs: add Zod 4 page

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2348 from hey-api/changeset-release/main

Version Packages

Daschi1 committed:
test: add valibot numberTypeToValibotSchema tests

Daschi1 committed:
feat(valibot): add integer format and BigInt handling and tests

Daschi1 committed:
test: update snapshots

mrlubos committed:
ci: fix pipelines

mrlubos committed:
Merge pull request #2353 from hey-api/ci/fix-fail

ci: fix pipelines

mrlubos committed:
Merge branch 'main' into issue-2194

Daschi1 committed:
test: switch integer-formats spec from JSON to YAML

Daschi1 committed:
test(valibot): switch spec from JSON to YAML

Daschi1 committed:
test(valibot): remove redundant assertion checks

Daschi1 committed:
refactor(valibot): extract number helper functions to separate file

Daschi1 committed:
test(valibot): remove redundant if conditions

Daschi1 committed:
test(valibot): remove redundant if conditions

mrlubos committed:
Create stale-starfishes-sell.md

mrlubos committed:
Merge pull request #2344 from Daschi1/issue-2194

github-actions[bot] committed:
Version Packages

mrlubos committed:
Merge pull request #2355 from hey-api/changeset-release/main

Version Packages

seriouslag committed:
refactor(nx-plugin): handle configs from builtin client packages

seriouslag committed:
patch(nx-plugin): removed generated tests

seriouslag committed:
test(nx-plugin): add experimental parser validation to configuration tests

seriouslag committed:
test(nx-plugin): add parser configuration for tests

seriouslag committed:
fix(openapi-client): update temp folder handling to use absolute path for directory creation

seriouslag committed:
fix(operation): ensure output case handling is robust by checking for object type

mrlubos committed:
Merge pull request #2356 from seriouslag/patch/add-output-object-check

mrlubos committed:
docs: add contributor manual overview

mrlubos committed:
Merge pull request #2358 from hey-api/docs/contributing-overview

docs: add contributor manual overview

mrlubos committed:
docs: fix Zod Mini shape API reference

mrlubos committed:
Merge pull request #2359 from hey-api/docs/zod-mini-shape

docs: fix Zod Mini shape API reference

Commit:
refactor: rename compiler export to tsc

mrlubos committed:
Merge pull request #2360 from hey-api/refactor/compiler-tsc

refactor: rename compiler export to tsc

mrlubos committed:
docs: add Effect and AdonisJS

mrlubos committed:
Merge pull request #2365 from hey-api/docs/adonis-effect

docs: add Effect and AdonisJS

seriouslag committed:
fix(nx-plugin): update command generation to correctly format plugin flags

seriouslag committed:
fix(openapi-client): update output path generation to use project root for generated files

seriouslag committed:
feat(nx-plugin): add projectReferences option

seriouslag committed:
feat(nx-plugin): copy generator templates to dist
